File: web/ax25/tests/listener-test-support.ts
/**
 * Test helpers for `Ax25Listener` unit tests. The {@link LoopbackTransport}
 * is the TS equivalent of the C# `LoopbackModem` — in-memory transport
 * with explicit inbound-injection + an observable list of outbound
 * frames.
 *
 * Kept in `tests/` (not `src/`) so consumers don't accidentally pull it
 * into production builds. New listener tests should import from this file
 * rather than re-rolling fixtures.
 */
import {
  type Ax25Frame,
  decodeFrame,
  encodeFrame,
} from "../src/frame.js";
import type { Ax25Transport } from "../src/transport.js";

/**
 * Block until `condition` returns `true`, polling every 20 ms. Throws
 * if the budget elapses. Mirrors `ListenerTestSupport.WaitFor` from C#.
 */
export async function waitFor(
  condition: () => boolean,
  budgetMs: number,
  reason?: string,
): Promise<void> {
  const deadline = Date.now() + budgetMs;
  while (Date.now() < deadline) {
    if (condition()) return;
    await new Promise((r) => setTimeout(r, 20));
  }
  throw new Error(
    `condition did not become true within ${budgetMs}ms${reason ? ` — ${reason}` : ""}`,
  );
}

/**
 * Tiny observable list — append + wait-for-count helper. Used by the
 * loopback transport's `sentFrames` so tests can block deterministically
 * on the outbound queue without polling sleeps littered through
 * assertions.
 */
export class ObservableList<T> {
  private readonly items: T[] = [];
  private readonly waiters: Array<{
    target: number;
    resolve: () => void;
    reject: (err: Error) => void;
    timer: ReturnType<typeof setTimeout>;
  }> = [];

  push(item: T): void {
    this.items.push(item);
    // Settle any waiters whose target has been reached.
    for (let i = this.waiters.length - 1; i >= 0; i--) {
      const w = this.waiters[i]!;
      if (this.items.length >= w.target) {
        clearTimeout(w.timer);
        this.waiters.splice(i, 1);
        w.resolve();
      }
    }
  }

  get count(): number {
    return this.items.length;
  }

  get(i: number): T {
    return this.items[i]!;
  }

  snapshot(): T[] {
    return this.items.slice();
  }

  async waitForCount(target: number, budgetMs: number): Promise<void> {
    if (this.items.length >= target) return;
    return new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => {
        const idx = this.waiters.findIndex((w) => w.resolve === resolve);
        if (idx !== -1) this.waiters.splice(idx, 1);
        reject(
          new Error(
            `only ${this.items.length}/${target} items after ${budgetMs}ms`,
          ),
        );
      }, budgetMs);
      this.waiters.push({ target, resolve, reject, timer });
    });
  }
}

/**
 * In-memory transport with explicit inbound injection. Mirrors the C#
 * `LoopbackModem` — what the listener calls `send` on appends to
 * `sentFrames`; what tests call `injectInbound` on shows up at the
 * listener's `onFrame` callback as decoded AX.25 bytes.
 */
export class LoopbackTransport implements Ax25Transport {
  private onFrame: ((bytes: Uint8Array) => void) | null = null;
  private startedFlag = false;
  /** Outbound AX.25 frames the listener has sent. */
  readonly sentFrames = new ObservableList<Uint8Array>();
  /** If true, outbound frames are counted but not appended to sentFrames. */
  dropOutbound = false;
  private outboundCountField = 0;

  async start(onFrame: (bytes: Uint8Array) => void): Promise<void> {
    this.onFrame = onFrame;
    this.startedFlag = true;
  }

  async send(bytes: Uint8Array): Promise<void> {
    this.outboundCountField++;
    if (this.dropOutbound) return;
    this.sentFrames.push(new Uint8Array(bytes));
  }

  async stop(): Promise<void> {
    this.startedFlag = false;
    this.onFrame = null;
  }

  /** Total send attempts regardless of `dropOutbound`. */
  get outboundCount(): number {
    return this.outboundCountField;
  }

  /** Inject a frame as if it had been received from the wire. */
  injectInbound(frame: Ax25Frame): void {
    if (!this.startedFlag || !this.onFrame) {
      throw new Error("transport not started");
    }
    const bytes = encodeFrame(frame);
    // Deliver synchronously — tests want deterministic ordering.
    this.onFrame(bytes);
  }

  /** Inject already-encoded AX.25 bytes (e.g. for malformed-frame tests). */
  injectInboundBytes(bytes: Uint8Array): void {
    if (!this.startedFlag || !this.onFrame) {
      throw new Error("transport not started");
    }
    this.onFrame(bytes);
  }

  /** Decode the n'th sent frame for assertions. */
  decodedSent(i: number): Ax25Frame {
    return decodeFrame(this.sentFrames.get(i));
  }
}

/** Add a budget to a promise — reject with TimeoutError if budget elapses. */
export async function withTimeout<T>(
  p: Promise<T>,
  budgetMs: number,
  label?: string,
): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const timer = setTimeout(
      () =>
        reject(
          new Error(
            `${label ?? "promise"} did not complete within ${budgetMs}ms`,
          ),
        ),
      budgetMs,
    );
    p.then(
      (v) => {
        clearTimeout(timer);
        resolve(v);
      },
      (err) => {
        clearTimeout(timer);
        reject(err);
      },
    );
  });
}

File: tests/Packet.Ax25.Tests/Session/ListenerTestSupport.cs
using System.Runtime.CompilerServices;
using System.Threading.Channels;
using Packet.Kiss;

namespace Packet.Ax25.Tests.Session;

/// <summary>
/// Shared test helpers for <c>Ax25Listener</c> unit tests. The
/// <see cref="LoopbackModem"/> in particular is used by every
/// <c>Ax25Listener*Tests.cs</c> file — extracted here so additions in one
/// file (e.g. a new injector overload) automatically benefit the others.
/// </summary>
/// <remarks>
/// Keep this internal to the test assembly. The Listener API doesn't
/// expose any kind of fake-modem fixture and shouldn't — the listener's
/// dependency on <see cref="IKissModem"/> is the seam tests poke through.
/// </remarks>
internal static class ListenerTestSupport
{
    /// <summary>
    /// Block until <paramref name="condition"/> returns <c>true</c>, polling
    /// every 20 ms. Throws <see cref="TimeoutException"/> if the deadline
    /// is reached first.
    /// </summary>
    public static async Task WaitFor(Func<bool> condition, TimeSpan budget, string? reason = null)
    {
        var deadline = DateTimeOffset.UtcNow + budget;
        while (DateTimeOffset.UtcNow < deadline)
        {
            if (condition()) return;
            await Task.Delay(20);
        }
        throw new TimeoutException($"condition did not become true within {budget}{(reason is null ? "" : $" — {reason}")}");
    }
}

/// <summary>
/// In-memory <see cref="IKissModem"/> whose inbound stream is a channel
/// the test writes <see cref="KissFrame"/>s into. Outbound
/// <c>SendFrameAsync</c> appends to <see cref="SentFrames"/> for the test
/// to assert against.
/// </summary>
/// <remarks>
/// Extended over the original (private nested in <c>Ax25ListenerTests</c>)
/// with these utilities the broader coverage tests need:
/// <list type="bullet">
/// <item><see cref="DropOutbound"/> — if <c>true</c>, outbound frames are
/// counted but not appended to <see cref="SentFrames"/>. Lets tests
/// simulate "peer never saw our UA" without changing the listener API.</item>
/// <item><see cref="SendDelay"/> — sleep this long inside
/// <see cref="SendFrameAsync"/> before recording. Lets tests probe whether
/// the listener's outbound path is on a separate thread from the inbound
/// pump (it should be — fire-and-forget).</item>
/// <item><see cref="OutboundFrameCount"/> — count of every send attempt
/// regardless of <see cref="DropOutbound"/>.</item>
/// </list>
/// </remarks>
internal sealed class LoopbackModem : IKissModem
{
    private readonly Channel<KissFrame> rx = Channel.CreateUnbounded<KissFrame>(new UnboundedChannelOptions
    {
        SingleReader = true,
        SingleWriter = false,
    });

    /// <summary>Outbound frames the listener / session has handed to the modem.</summary>
    public ObservableList<ReadOnlyMemory<byte>> SentFrames { get; } = new();

    /// <summary>If <c>true</c>, outbound frames are dropped — <see cref="SentFrames"/> stays empty.</summary>
    public bool DropOutbound { get; set; }

    /// <summary>If non-zero, <see cref="SendFrameAsync"/> sleeps for this long before recording.</summary>
    public TimeSpan SendDelay { get; set; } = TimeSpan.Zero;

    private int outboundCount;

    /// <summary>Total send attempts regardless of <see cref="DropOutbound"/>.</summary>
    public int OutboundFrameCount => Volatile.Read(ref outboundCount);

    /// <summary>Push an inbound AX.25 frame as if it had been heard off the air.</summary>
    public void InjectInbound(Ax25Frame frame)
    {
        var kf = new KissFrame((byte)0, KissCommand.Data, frame.ToBytes().ToArray());
        rx.Writer.TryWrite(kf);
    }

    /// <summary>
    /// Push an inbound raw KISS frame. Useful for non-Data commands the
    /// listener should skip (TX echoes etc.).
    /// </summary>
    public void InjectInboundRaw(KissFrame frame)
    {
        rx.Writer.TryWrite(frame);
    }

    public async Task SendFrameAsync(ReadOnlyMemory<byte> ax25Bytes, CancellationToken cancellationToken = default)
    {
        Interlocked.Increment(ref outboundCount);
        if (SendDelay > TimeSpan.Zero)
        {
            await Task.Delay(SendDelay, cancellationToken).ConfigureAwait(false);
        }
        if (!DropOutbound)
        {
            SentFrames.Add(ax25Bytes);
        }
    }

    public async IAsyncEnumerable<KissFrame> ReadFramesAsync(
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        while (await rx.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
        {
            while (rx.Reader.TryRead(out var f))
            {
                yield return f;
            }
        }
    }

    /// <summary>
    /// When non-null, <see cref="SendFrameWithAckAsync"/> is supported: the frame
    /// is recorded like a plain send and the call's receipt resolves when the test
    /// releases this gate (one release = one TX-completion echo) — letting tests
    /// control exactly when a frame "clears the air". When null (default), ACKMODE
    /// is unsupported and the call throws, like a plain non-ACKMODE modem.
    /// </summary>
    public SemaphoreSlim? AckEchoGate { get; set; }

    public async Task<AckModeReceipt> SendFrameWithAckAsync(
        ReadOnlyMemory<byte> ax25Bytes, TimeSpan? timeout = null, ushort? sequenceTag = null,
        CancellationToken cancellationToken = default)
    {
        if (AckEchoGate is not { } gate)
        {
            throw new NotSupportedException("this LoopbackModem has no ACKMODE (AckEchoGate not set).");
        }
        var queued = DateTimeOffset.UtcNow;
        Interlocked.Increment(ref outboundCount);
        if (!DropOutbound)
        {
            SentFrames.Add(ax25Bytes.ToArray());
        }
        await gate.WaitAsync(cancellationToken).ConfigureAwait(false);
        return new AckModeReceipt(sequenceTag ?? 0, queued, DateTimeOffset.UtcNow);
    }
    public Task SetTxDelayAsync(byte v, CancellationToken c = default) => Task.CompletedTask;
    public Task SetPersistenceAsync(byte v, CancellationToken c = default) => Task.CompletedTask;
    public Task SetSlotTimeAsync(byte v, CancellationToken c = default) => Task.CompletedTask;
    public Task SetTxTailAsync(byte v, CancellationToken c = default) => Task.CompletedTask;
}

/// <summary>
/// Tiny thread-safe list with a "wait until count reaches N" helper.
/// Lets tests block deterministically on the modem's outbound queue
/// without polling sleeps littered through the assertions.
/// </summary>
internal sealed class ObservableList<T>
{
    private readonly List<T> items = new();
    private readonly object gate = new();
    private readonly List<TaskCompletionSource<bool>> waiters = new();

    public void Add(T item)
    {
        List<TaskCompletionSource<bool>> toComplete;
        lock (gate)
        {
            items.Add(item);
            toComplete = waiters.ToList();
            waiters.Clear();
        }
        foreach (var w in toComplete) w.TrySetResult(true);
    }

    public int Count
    {
        get { lock (gate) return items.Count; }
    }

    public T this[int i]
    {
        get { lock (gate) return items[i]; }
    }

    public List<T> SnapshotList()
    {
        lock (gate) return items.ToList();
    }

    public async Task WaitForCountAsync(int target, TimeSpan budget)
    {
        var deadline = DateTimeOffset.UtcNow + budget;
        while (true)
        {
            Task wait;
            lock (gate)
            {
                if (items.Count >= target) return;
                var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
                waiters.Add(tcs);
                wait = tcs.Task;
            }
            var remaining = deadline - DateTimeOffset.UtcNow;
            if (remaining <= TimeSpan.Zero) throw new TimeoutException($"only {Count}/{target} items after {budget}");
            var done = await Task.WhenAny(wait, Task.Delay(remaining));
            if (done != wait) throw new TimeoutException($"only {Count}/{target} items after {budget}");
        }
    }
}

internal static class ListenerTestTaskExtensions
{
    public static async Task<T> WithTimeout<T>(this Task<T> task, TimeSpan budget)
    {
        var done = await Task.WhenAny(task, Task.Delay(budget));
        if (done != task) throw new TimeoutException($"task did not complete within {budget}");
        return await task;
    }

    public static async Task WithTimeout(this Task task, TimeSpan budget)
    {
        var done = await Task.WhenAny(task, Task.Delay(budget));
        if (done != task) throw new TimeoutException($"task did not complete within {budget}");
        await task;
    }
}
